doc: introduce temporary environment and shared environment (#2987)
JuliaLang/Pkg.jl: docs/src/environments.md

Example = "7876af07-990d-54b4-ab0e-23690620f79a"

julia> print(read("Manifest.toml", String))
# This file is machine-generated - editing it directly is not advised

[[Example]]
git-tree-sha1 = "46e44e869b4d90b96bd8ed1fdcf32244fddfb6cc"
uuid = "7876af07-990d-54b4-ab0e-23690620f79a"
version = "0.5.3"
```

> This new environment is completely separate from the one we used earlier. See [`Project.toml` and `Manifest.toml`](@ref Project-and-Manifest) for more detailed explaination on them.

Reviewer: ```suggestion
This new environment is completely separate from the one we used earlier. See [`Project.toml` and `Manifest.toml`](@ref Project-and-Manifest) for a more detailed explanation.
```

    ```bash
    $ julia --project=. myscript.jl
    ```

## Temporary environments

> It is not uncommon to test some package without installing it into your usual environment.

Reviewer 2: ```suggestion
Temporary environments make it easy to start an environment from a blank slate to test a package or set of
packages, and have Pkg automatically delete the environment when you're done.
```

Perhaps when troubleshooting an especially thorny package management issue, you may want to ensure the
temporary environment's package depot is completely independent from any other package environment by
setting the [`JULIA_DEPOT_PATH`](https://docs.julialang.org/en/v1/manual/environment-variables/#JULIA_DEPOT_PATH) environment variable.
For example,

```console
# Linux Bash
root@dc76efc35971:/# JULIA_DEPOT_PATH=/tmp/juliatmp julia

(@v1.6) pkg> st
  Installing known registries into `/tmp/juliatmp`
      Status `/tmp/juliatmp/environments/v1.6/Project.toml` (empty project)

(@v1.6) pkg> add Example
    Updating registry at `/tmp/juliatmp/registries/General.toml`
   Resolving package versions...
   Installed Example ─ v0.5.3
    Updating `/tmp/juliatmp/environments/v1.6/Project.toml`
  [7876af07] + Example v0.5.3
    Updating `/tmp/juliatmp/environments/v1.6/Manifest.toml`
  [7876af07] + Example v0.5.3
Precompiling project...
  1 dependency successfully precompiled in 0 seconds
```

This way, anything downloaded or modified by package operations will be stored in the temporary directory
`/tmp/juliatmp`, with no need to worry that it might break your regular workflow.

> ## [shared environment](@id shared-environment)

Reviewer 3: I don't think this description is correct for shared environments. A shared environment is just an environment that can be referred to by name and is put in `~/.julia/environments/`. For example:

```
(@v1.7) pkg> activate --shared Foo
  Activating project at `~/.julia/environments/Foo`
```

This is completely separate from `LOAD_PATH` and a shared environment created like above will not be in the default load path if you activate a different project.

```

Loading `BenchmarkTools` in MyProject environment works even if this package is not recorded in `~/MyProject/Project.toml`. This is
because Julia maintains a environments stack via the `LOAD_PATH` variable, see also the [environment stacks](https://docs.julialang.org/en/v1/manual/code-loading/#Environment-stacks) section in the julia manual.

```julia-repl
julia> LOAD_PATH
3-element Vector{String}:
 "@"
 "@v#.#"
 "@stdlib"
```

Briefly speaking, Julia will try to load package `BenchmarkTools` in `@` environment, which is the current activated project.
If not found then try the version-specific global environment `@v#.#`, which is stored in `~/.julia/environments/v1.6` for Julia 1.6.
In the previous case, Julia found `BenchmarkTools` in `@v#.#`, i.e., `~/.julia/environments/v1.6/Project.toml`,
thus it successfully loads the package.

A convenient command to create custom shared environment is to prefix the name with `@`:

```julia-repl
(@v1.6) pkg> activate @mytoolbox # require Julia 1.4 or later
  Activating new project at `~/.julia/environments/mytoolbox`
```

This provides a quick way to activate the project environment stored in `~/.julia/environments` folder.

> ### Shared environment caveat: the reproducibility issue
> 
> Using shared environment can cause reproducibility issues. For instance,

Reviewer: ```suggestion
!!! warning Shared environments and reproducibility
    Because shared environments allow you to load packages without specifying them in a particular environment's
    `Project.toml` and `Manifest.toml`, you may encounter problems with reproducibility.
```

Author: I'm not sure if it's appropriate to use `!!! ` for a long thread like this, so I kept the `###` title here.

```julia
(MyProject) pkg> generate MyPackage
  Generating  project MyPackage:
    MyPackage/Project.toml
    MyPackage/src/MyPackage.jl

# ... some modifications to MyPackage/src/MyPackage.jl

julia> print(read("MyPackage/Project.toml", String))
name = "MyPackage"
uuid = "c10752ec-72c1-4eb2-a399-580941ff7382"
version = "0.1.0"

julia> print(read("MyPackage/src/MyPackage.jl", String))
module MyPackage

using BenchmarkTools

greet() = print("Hello World!")

end # module
```

> Now when you tries to load `MyPackage`, because `BenchmarkTools` is recorded in the global shared environment `v#.#`,
> Julia still succeeds loading it but will print a warning message:

Reviewer: ```suggestion
When you try to load `MyPackage`, Julia searches for and finds `BenchmarkTools` in the global shared environment `v#.#`, so loading is successful, but a warning message is printed:
```

```julia
julia> using MyPackage
┌ Warning: Package MyPackage does not have BenchmarkTools in its dependencies:
│ - If you have MyPackage checked out for development and have
│   added BenchmarkTools as a dependency but haven't updated your primary
│   environment's manifest file, try `Pkg.resolve()`.
│ - Otherwise you may need to report an issue with MyPackage
└ Loading BenchmarkTools into MyPackage from project dependency, future warnings for MyPackage are suppressed.
```

> Imaging that you want to ship your wonderful package to other people, because `BenchmarkTools` is not recoreded in `MyPackage/Project.toml`,
> you can't ensure that `BenchmarkTools` can be successfully loaded even if they initialize your package with `pkg> instantiate`.
> In such sense, this is a reproducibility warning; when you develop a package that meant to be shared by others, you should avoid doing this.

Reviewer: ```suggestion
Imagine that you want to ship your wonderful package to other people. Because `BenchmarkTools` is not recorded in `MyPackage/Project.toml`,
you can't ensure that `BenchmarkTools` will be successfully loaded. For this reason, developers of shared packages should be especially mindful of reproducibility issues when using shared environments.
```

    ```bash
    $ julia --project=. myscript.jl
    ```

## Temporary environment

It is not uncommon to test some package without installing it into your usual environment. For instance,
when you try to write a bug report with a consistent way to reproduce, or you want to try out a very
old/new package version that is incompatible with your current environment.

```julia-repl
(@v1.6) pkg> activate --temp # require Julia 1.5 or later
  Activating new environment at `/var/folders/34/km3mmt5930gc4pzq1d08jvjw0000gn/T/jl_a31egx/Project.toml`

(jl_a31egx) pkg> add Example
    Updating registry at `~/.julia/registries/General`
   Resolving package versions...
    Updating `/private/var/folders/34/km3mmt5930gc4pzq1d08jvjw0000gn/T/jl_a31egx/Project.toml`
  [7876af07] + Example v0.5.3
    Updating `/private/var/folders/34/km3mmt5930gc4pzq1d08jvjw0000gn/T/jl_a31egx/Manifest.toml`
  [7876af07] + Example v0.5.3
```

The temporary directory will be removed when you exit Julia so you don't need to worry this operation
breaks your common workflow.

> Note that this still reuses the contents in the global depot path in `~/.julia`. Thus you might still hit

Reviewer 3: I felt this started to get way too long. I have almost never needed to change the depot path because things in it are content tracked so it should never " break your common workflow.". We have to try keep the text a bit short so readers don't get fatigued.
